origin-mock a flexible mock server

ejemba opened the pull request:
This is a mock server expected to be used in advanced tests. 
It can be extended, but already contains useful primitives. See test/mock/README.md
It is intended to be included in pod definition as "openshift/mock"

The whole thing is in test/mock directory.

- .gitignore added in mock/bin
- Dockerfile, server.go added
- `make release` updated.
- Chaining concept of call.
- put/get/contains/delete
- suicide command (delayed suicide possible)

pmorie commented on test/mock/README.md:
```diff
@@ -0,0 +1,162 @@
+
+This kills the server itself immediatly.
+
+## /suicide/{delay}
```
I like this feature but I'm not sure about the name.  How about 'selfdestruct' instead?

ironcladlou replied:
Or just "shutdown"?

ejemba replied:
@pmorie Yeah, the name can be somehow weird. 
@ironcladlou 'shutdown' is a good candidate.

pmorie commented:
@ejemba This is cool - do you have a test that uses it?  It would help to review the whole thing w/ a test that applies it.

ironcladlou commented:
Forgive me if this is obvious, but I'm a bit surprised that this works as a `FROM scratch` image since it's using `net/http`, which requires dynamic linking. What am I missing?

ejemba commented:
@pmorie in fact, It will be easier if the mock goes in the master, but I can branch my PR and write a test that use it to give some example.

ejemba commented:
@ironcladlou have a look at this
https://github.com/openshift/origin/blob/master/hack/build-cross.sh#L16-L20
the build process in which images like hello-openshift or origin/mock are involved includes flags to make them static especially for this purpose.

bparees commented:
"mock" as a name is a little generic for my taste (particularly "test/mock"), how about test/mock-app or test/mock-app-server ?  (I know we discussed this a bit in email too, but with no solid conclusion)

pmorie commented:
Mock-api-server?
On Sun, Apr 26, 2015 at 11:22 PM Ben Parees <notifications@github.com>
wrote:

> "mock" as a name is a little generic for my taste (particularly
> "test/mock"), how about test/mock-app or test/mock-app-server ? (I know we
> discussed this a bit in email too, but with no solid conclusion)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1903#issuecomment-96486229>.
>

bparees commented:
@pmorie it's not actually mocking the openshift api server, right?  so i think that would be confusing.

pmorie commented:
@bparees I suppose that's true. How about mock-http-server or mock-rest?
On Sun, Apr 26, 2015 at 11:34 PM Ben Parees <notifications@github.com>
wrote:

> @pmorie <https://github.com/pmorie> it's not actually mocking the
> openshift api server, right? so i think that would be confusing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1903#issuecomment-96488664>.
>

ejemba commented:
@bparees I agree. I like test/mock-app because, this actually is what this server is : a mock of an application. 
@pmorie I think test/mock-rest or test/mock-http-server more describes a mean than a purpose.

mfojtik commented on test/mock/server.go:
```diff
@@ -0,0 +1,253 @@
+	}
+
+	// port of the mock server
+	port = os.Getenv("MOCK_PORT")
```
there is an util method for returning env vars / assigning default values to them

mfojtik commented on test/mock/server.go:
```diff
@@ -0,0 +1,253 @@
+	"github.com/gorilla/mux"
+)
+
+var name string
```
you can group these vars:

```
var (
...
)
```

or make a struct Server

ejemba replied:
true. will do it.

mfojtik commented on test/mock/server.go:
```diff
@@ -0,0 +1,253 @@
+	delayStr := p["delay"]
+
+	delay, err := strconv.Atoi(delayStr)
+
```
nuke newlines like this (the if is related to the statement above)

mfojtik commented on test/mock/server.go:
```diff
@@ -0,0 +1,253 @@
+	}
+
+	timer := time.NewTimer(time.Second * time.Duration(delay))
+	fmt.Printf("%s will autokill in %d second(s)\n", name, delay)
```
Since we already have glog in Godeps, I would use glog to make the output consistent and have timestamps.

ejemba replied:
Ok. will do it.

ejemba commented:
I updated it all. 
I call the image : openshift/mock-app
I refactor the code following @mfojtik advise. 
Added
- /env/{var}
- /hostname
- /shutdown and /shutdown/{delay} replace /suicide

bparees commented:
i'd still like to see the actual test directory name updated.

ejemba commented:
@bparees mock directory has been renamed to mock-app as well. 
/hostname has been added.

bparees commented:
lgtm.  @pmorie @ironcladlou @smarterclayton any other comments?

ironcladlou commented:
My only remaining comment is I'd like to see some meaningful test that uses the feature in order to justify maintaining it. Even if it's in another PR somewhere.

ejemba commented:
@ironcladlou you're right. I was writing some tests in another branch. I keep you informed.

danmcp commented:
@ejemba Any updates?  Is this pull still valid?

ejemba commented:
@danmcp Yes it is still valid. I'm working on the second PR from this one, to show some tests using the mock-app, I already manage to easily produce intra pod communication assertions, inter pods is coming, in the second update I'll show Assertions on DeploymentConfig  . 
From time to time e2e broke, I had some issues making tests working, but now it's fine.
I'll create the second PR from this one today or tomorrow. 
But all this is depending on the fix of making static go binary that I'm also addressing on a third PR. I keep you informed.

ejemba commented:
Created https://github.com/openshift/origin/pull/2999 to illustrate how we use the mock-app

@ironcladlou @bparees @pmorie @smarterclayton. 


sorry for the delay, had a lot to understand regarding the whole openshift objects. 
Plus they move constantly

smarterclayton commented on test/mock-app/.gitignore:
```diff
@@ -0,0 +1 @@
+mock-app
```
I would still keep this directory under images - `images/test/mock-app`

ejemba replied:
@smarterclayton I do that.

danmcp commented:
Bump

ejemba commented:
I've just added the 
  /fail  and
  /fail/{exitCode}
so the server can exit with a specific exitCode by calling 
*http://server:port/fail* or *http://server:port/fail/127*

The docker image has been moved in images/test/mock 

The image can be used as long as Travis is passing.

danmcp commented:
@ejemba Bump

danmcp commented:
Closing due to age without prejudice.
